File: app/src/main/java/com/denproj/posmanongjaks/viewModel/SalesFragmentViewmodel.java
package com.denproj.posmanongjaks.viewModel;

import android.util.Log;

import androidx.databinding.ObservableField;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.denproj.posmanongjaks.hilt.qualifier.FirebaseImpl;
import com.denproj.posmanongjaks.model.Item;
import com.denproj.posmanongjaks.model.Product;
import com.denproj.posmanongjaks.repository.base.AddOnsRepository;
import com.denproj.posmanongjaks.repository.base.ProductRepository;
import com.denproj.posmanongjaks.util.OnDataReceived;
import com.denproj.posmanongjaks.util.OnFetchFailed;
import com.denproj.posmanongjaks.util.OnUpdateUI;

import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class SalesFragmentViewmodel extends ViewModel {
    ProductRepository productRepository;
    AddOnsRepository addOnsRepository;

    @Inject
    public SalesFragmentViewmodel(
            @FirebaseImpl ProductRepository productRepository,
            @FirebaseImpl AddOnsRepository addOnsRepository) {
        this.productRepository = productRepository;
        this.addOnsRepository = addOnsRepository;
    }

    public void loadProductsOfBranch(String branchId, OnUpdateUI<List<Product>> listOnUpdateUI) {
        productRepository.fetchProductsFromBranch(branchId)
                .thenAccept(listOnUpdateUI::onSuccess)
                .exceptionally(throwable -> {
                    listOnUpdateUI.onFail(new Exception(throwable));
                    return null;
                });
    }

    public MutableLiveData<List<Product>> observeProductListOfBranch(String branchId, OnFetchFailed onFetchFailed) {
        return productRepository.observeProductList(branchId, onFetchFailed);
    }

    public MutableLiveData<List<Item>> observeAddOnsListOfBranch(String branchId, OnFetchFailed onFetchFailed) {
        return addOnsRepository.observeAddOnsList(branchId, onFetchFailed);
    }

    public void loadAddOns(String branchId, OnUpdateUI<List<Item>> onUpdateUI) {
        addOnsRepository.getAddOnsRepository(branchId, new OnDataReceived<List<Item>>() {
            @Override
            public void onSuccess(List<Item> result) {
                onUpdateUI.onSuccess(result);
            }

            @Override
            public void onFail(Exception e) {
                onUpdateUI.onFail(e);
            }
        });
    }

    public ProductRepository getProductRepository() {
        return productRepository;
    }

    public void setProductRepository(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    public AddOnsRepository getAddOnsRepository() {
        return addOnsRepository;
    }

    public void setAddOnsRepository(AddOnsRepository addOnsRepository) {
        this.addOnsRepository = addOnsRepository;
    }
}

File: app/src/main/java/com/denproj/posmanongjaks/viewModel/SalesFragmentViewmodel.kt
package com.denproj.posmanongjaks.viewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.denproj.posmanongjaks.hilt.qualifier.FirebaseImpl
import com.denproj.posmanongjaks.hilt.qualifier.FirestoreImpl
import com.denproj.posmanongjaks.model.Item
import com.denproj.posmanongjaks.model.Product
import com.denproj.posmanongjaks.repository.base.AddOnsRepository
import com.denproj.posmanongjaks.repository.base.ProductRepository
import com.denproj.posmanongjaks.repository.impl.SessionRepository
import com.denproj.posmanongjaks.util.OnDataReceived
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SalesFragmentViewmodel @Inject constructor(
    @FirestoreImpl var productRepository: ProductRepository,
    @FirestoreImpl var addOnsRepository: AddOnsRepository,
    var sessionRepository: SessionRepository
) : ViewModel() {

    private val _items: MutableLiveData<List<Item?>?> = MutableLiveData(emptyList())
    val items: LiveData<List<Item?>?> = _items

    private val _products: MutableLiveData<List<Product>> = MutableLiveData(null)
    val products: LiveData<List<Product>> = _products

    private val _product: MutableLiveData<Product> = MutableLiveData(null)
    val product: LiveData<Product> = _product

    private val _errors: MutableLiveData<Exception> = MutableLiveData(null)
    val errors: LiveData<Exception> = _errors


    init {
        viewModelScope.launch {
            var sessionSimple = sessionRepository.getSession()
            if (sessionSimple != null) {
                addOnsRepository.observeAddOnsList(sessionSimple.branchId!!, object : OnDataReceived<List<Item?>?> {
                    override fun onSuccess(data: List<Item?>?) {
                        if (data != null) {
                            _items.postValue(data)
                        }
                    }

                    override fun onFail(e: Exception?) {
                        _errors.postValue(e ?: Exception("An error occurred when loading add ons."))
                    }
                })
                productRepository.observeProductList(sessionSimple.branchId, object : OnDataReceived<List<Product>?> {
                    override fun onSuccess(data: List<Product>?) {
                        _products.postValue(data ?: emptyList())
                    }

                    override fun onFail(e: Exception?) {
                        _errors.postValue(e ?: Exception("An error occurred when loading products ons."))
                    }
                })
            }
        }
    }
}
